### PR TITLE
Implement LayerNorm

### PR DESCRIPTION
**Pull request by daemyung**
## 🙏 Describe the pull request

LayerNorm is implemented.

## ✅ Checklist

- [x] Code follows the project's coding conventions and style.
- [x] Tests have been added or updated to cover the changes.
- [x] Documentation has been updated, if necessary.

### CODE REVIEW

**Comment by daemyung** [comment]
역전파 구현 중입니다.

**Comment by daemyung** [comment]
@ansteve @mejai1206 리뷰 부탁드립니다.

File: trident/operation/layer_norm.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+class LayerNorm(torch.autograd.Function):
+    @staticmethod
+    def forward(*args, **kwargs):
+        return LayerNorm.__forward(*args, **kwargs)
+
+    @staticmethod
+    def setup_context(ctx, inputs, output):
+        inp, norm_sh, wgt, bis, eps, mean, var = inputs
+        ctx.save_for_backward(inp, wgt, bis, mean, var)
```
Comment (kakao-steve-ai) [inline trident/operation/layer_norm.py]: std를 var로 오타치신 거 같습니다.

File: trident/operation/layer_norm.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        vec_sz = LayerNorm.__get_vec_sz(norm_sh)
+        num_vec = inp.numel() // vec_sz
+
+        def grid(meta):
+            return [num_vec, ]
+
+        out = torch.empty_like(inp)
+        vec_bs = triton.next_power_of_2(vec_sz)
+
+        kernel.LayerNorm.forward[grid](inp.view(num_vec, vec_sz), vec_sz, vec_sz, wgt, bis, eps, mean, std,
```
Comment (kakao-steve-ai) [inline trident/operation/layer_norm.py]: vector size와 vector stride에 모두 vec_sz를 사용하는데, 두 변수를 분리해놓은 이유가 있을까요? 나중에 더 높은 dimension 지원을 위해 미리 분리해놓은걸까요?

File: trident/kernel/layer_norm.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+from trident import language
+
+
+class LayerNorm:
+    @staticmethod
+    @triton.jit
+    def forward(inp_ptr, vec_sz, vec_st, wgt_ptr, bis_ptr, eps, mean_ptr, std_ptr, out_ptr,
+                vec_bs: triton.language.constexpr):
+        pid = triton.language.program_id(0)
+        pid = triton.language.program_id(0)
```
Comment (kakao-steve-ai) [inline trident/kernel/layer_norm.py]: 오타로 같은게 두줄 있습니다.

File: trident/operation/layer_norm.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    def backward(ctx, *grad_outputs):
+        return LayerNorm.__backward(*grad_outputs, *ctx.saved_tensors, ctx.norm_sh)
+
+    @staticmethod
+    def __forward(inp, norm_sh, wgt, bis, eps, mean, std):
+        vec_sz = LayerNorm.__get_vec_sz(norm_sh)
+        num_vec = inp.numel() // vec_sz
+
+        def grid(meta):
+            return [num_vec, ]
```
Comment (mejai1206) [inline trident/operation/layer_norm.py]: 이 부분은 튜플인걸 자주 봤는데, 이번엔 리스트인 이유가 있을까요?

File: trident/kernel/layer_norm.py (hunk 3 above)
Comment (mejai1206) [inline trident/kernel/layer_norm.py]: Pid 가 중복 할당됩니다.

File: trident/kernel/layer_norm.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        wgt_blk = triton.language.arange(0, vec_bs)
+        inp_blk = wgt_blk + pid * vec_st
+        msk = wgt_blk < vec_sz
+
+        inp = triton.language.load(inp_ptr + inp_blk, msk, 0)
```
Comment (mejai1206) [inline trident/kernel/layer_norm.py:30]: mean, std, var와 같은 값들은 전체에서 집계하는 것이 아니라
블럭단위로 집계해도 괜찮을까요??

File: trident/operation/layer_norm.py (hunk 2 above)
Comment (daemyung) [reply trident/operation/layer_norm.py]: `stride`와 `size`가 다를 수 있는 상황을 가정해서 만들긴 했는데, 삭제하도록 하겠습니다,

File: trident/operation/layer_norm.py (hunk 4 above)
Comment (daemyung) [reply trident/operation/layer_norm.py]: 없습니다. 두 개를 구분하는게 좋을까요?

File: trident/kernel/layer_norm.py (hunk 5 above)
Comment (daemyung) [reply trident/kernel/layer_norm.py:30]: 블럭의 크기가 벡터입니다. 그러므로 정규화하려는 단위인 블록의 크기로 `mean`, `std`, `var`를 계산하는게 올바릅니다.

File: trident/operation/layer_norm.py (hunk 4 above)
Comment (mejai1206) [reply trident/operation/layer_norm.py]: 아 단지 궁금해서 여쭤보았습니다.
동작에 문제가 없다면 상관 없을 것 같습니다.

File: trident/operation/layer_norm.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    @staticmethod
+    def __forward(inp, norm_sh, wgt, bis, eps, mean, std):
+        vec_sz = LayerNorm.__get_vec_sz(norm_sh)
+        num_vec = inp.numel() // vec_sz
+
+        def grid(meta):
+            return (num_vec,)
+
+        out = torch.empty_like(inp)
+        vec_bs = triton.next_power_of_2(vec_sz)
```
Comment (kakao-steve-ai) [inline trident/operation/layer_norm.py:47]: vector size를 블록 사이즈로 사용하고 있어서, vector size가 1024를 넘어가면 한 블록이 너무 커져서 속도가 느려질 거 같습니다. 이 부분 최적화는 다른 PR로 진행해도 좋을 거 같습니다.

Comment (daemyung) [reply trident/operation/layer_norm.py:47]: 그렇지 않습니다. GPGPU 프로그래밍을 할 때 고려해야할 점을 하지 않도록 하는 것이 Triton의 장점입니다. [LayerNorm의 성능](https://github.com/kakaobrain/trident/wiki/Performance#layernorm) 확인해보면 알 수 있습니다.

File: trident/operation/layer_norm.py (hunk 4 above)
Comment (daemyung) [reply trident/operation/layer_norm.py]: 튜플과 리스트의 차이에 대해서 좀 더 알아봅시다.

File: trident/operation/layer_norm.py (hunk 6 above)
Comment (kakao-steve-ai) [reply trident/operation/layer_norm.py:47]: 스크린샷 2023-07-05 오후 1 42 41

제 V100 장비에서 LayerNorm 성능 돌려봤을때 이렇게 나오는데, 위키에 있는 거랑 많이 다르게 나오네요. 원인이 뭘까요?

Comment (daemyung) [reply trident/operation/layer_norm.py:47]: 최신 Triton 설치 해보시겠어요? Wiki에 있는 그림도 V100 결과입니다.

File: trident/operation/layer_norm.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        grad_bis = None if bis is None else torch.zeros_like(bis)
+        vec_bs = triton.next_power_of_2(vec_sz)
+
+        # TODO: Create a tensor in a kernel after a bug of Triton is fixed.
+        wgt = torch.zeros(vec_sz, device='cuda').fill_(1) if wgt is None else wgt
+
+        kernel.LayerNorm.backward[grid](grad_out, inp, grad_inp, vec_sz, wgt, grad_wgt, grad_bis, mean, std,
+                                        vec_bs=vec_bs)
+
+        return grad_inp, None, grad_wgt, grad_bis, None, None, None, None,
```
Comment (mejai1206) [inline trident/operation/layer_norm.py:73]: 각 None들이 위치 하는 곳이 어떤 의미인지 궁금합니다.

File: trident/operation/layer_norm.py (hunk 6 above)
Comment (daemyung) [reply trident/operation/layer_norm.py:47]: 음,,, 지금 돌려보니 벡터의 크기가 32768 이상인 경우부터 성능이 느려지네요. 분석이 필요해 보입니다.

File: trident/operation/layer_norm.py (hunk 7 above)
Comment (daemyung) [reply trident/operation/layer_norm.py:73]: 입력에 해당하는 `gradient`입니다. `LayerNorm`의 `forward` 입력이 `inp, norm_sh, wgt, bis, eps, mean, std`이고 이 입력들의 gradient를 `grad_inp, None, grad_wgt, grad_bis, None, None, None, None`로 반환합니다.

File: trident/operation/layer_norm.py (hunk 6 above)
Comment (kakao-steve-ai) [reply trident/operation/layer_norm.py:47]: triton.autotune 이용해서 블록 사이즈 여러개 config로 넣어서 해보면 최적값을 구할 수 있을 것 같습니다.

Comment (daemyung) [reply trident/operation/layer_norm.py:47]: 블록 사이즈가 바뀌면 커널이 바뀌어야 합니다. 추후에 개선하도록 하겠습니다.

**Review by kakao-steve-ai — APPROVED** [review]
좋아보입니다.
